### ISSUE TITLE
Metal OOM (>72 GB single-buffer) on text-only prompts >~34K tokens with hybrid SSM models (Qwen3.5)

### ISSUE THREAD

**Issue by st-adam** [jjang-ai/vmlx#89]
## Environment

- Hardware: Apple M3 Max, 128 GB unified memory
- OS: macOS (Metal single-buffer cap ~72 GiB / 77,309,411,328 bytes)
- vmlx: 1.3.61
- mlx-lm: 0.31.2
- Model: `mlx-community/Huihui-Qwen3.5-27B-Claude-4.6-Opus-abliterated-4bit` (hybrid: GatedDeltaNet + attention, 48 layers)
- Server flags include `--use-paged-cache --continuous-batching --prefill-step-size 512 --prefill-batch-size 1 --kv-cache-quantization q8`

## Bug

Long text-only prompts (>~34K tokens) crash with:

```
libc++abi: terminating due to uncaught exception of type std::runtime_error:
[METAL] Command buffer execution failed: Insufficient Memory
(00000008:kIOGPUCommandBufferCallbackErrorOutOfMemory)
```

A single `attention_scores` allocation of `(1, num_heads, seq_len, seq_len) * 2 bytes` (fp16) at seq_len ≈ 48 000 is ~147 GiB, which exceeds Metal's per-buffer cap of ~72 GiB and OOMs instantly — even though total unified memory is 128 GiB.

## Root cause

`vmlx_engine/mllm_batch_generator.py` unconditionally guards hybrid SSM models against chunked prefill:

- L1428 (fast path): `if not has_images and not self._is_hybrid:` — forces hybrid models through the full model (`self.model(...)`) rather than `language_model(...)` with chunked prefill
- L1444 (chunked path): `if not has_images and seq_len > self.prefill_step_size * 2 and not self._is_hybrid:` — same guard blocks chunked prefill for long prompts

Both paths end up computing full-sequence attention scores for hybrid models, which is what triggers the single-buffer OOM.

## Safety analysis (Qwen3.5)

Chunked prefill is safe for Qwen3.5 hybrid (GatedDeltaNet + attention):

- `mlx_lm/models/qwen3_5.py` uses `create_attention_mask(h, cache[fa_idx])` and `create_ssm_mask(h, cache[ssm_idx])` — both delegate to `cache.make_mask(N)`, which is cache-aware and produces correct masks for any chunk size.
- Both `KVCache` and `ArraysCache` (SSM recurrent state) expose proper `state` properties that carry across chunks.
- vmlx already calls `mx.eval(cache states)` between chunks, materializing KV pairs and SSM recurrent state before the next chunk runs.

Other hybrid architectures have not been verified in mlx_lm's mask creation code, so removing the guards unconditionally is too aggressive.

## Repro

1. Serve Qwen3.5 27B via `vmlx serve ... --prefill-step-size 512 --use-paged-cache --continuous-batching` on M3 Max 128 GB.
2. Send a single request with a ~48K-token text-only prompt.
3. Server crashes with the Metal OOM above, before generating any tokens.

## Proposed fix

Opt-in env var `VMLX_ALLOW_HYBRID_CHUNKED_PREFILL=1` that lets hybrid models take the chunked prefill path on text-only requests. Default remains OFF — no behavior change for existing users.

PR: #<fill-in-after>

Happy to expand verification or adjust the gating strategy (e.g. model-name allowlist, architecture introspection) if preferred.

**Comment by jjang-ai** [comment]
Thanks for the excellent root-cause analysis + safety review — fix landed exactly along the lines you proposed.

**Fix** (`vmlx_engine/mllm_batch_generator.py` L1463–L1500, on main):

Opt-in env var `VMLX_ALLOW_HYBRID_CHUNKED_PREFILL=1` relaxes both guard sites:

- L1428 / L1444 still default to the full-model path for hybrid models (conservative — other hybrid architectures in mlx-lm haven't had their `create_*_mask` paths audited for chunked prefill correctness).
- When `VMLX_ALLOW_HYBRID_CHUNKED_PREFILL=1` is set, hybrid text-only requests take the chunked `language_model(...)` path with `prefill-step-size`-sized chunks, matching the non-hybrid text-only flow.
- Between chunks, we `mx.eval` on cache states (KV + SSM recurrent state) so the next chunk starts from materialized tensors — as you noted was already the case for the non-hybrid path.

Qwen3.5 hybrid (GatedDeltaNet + attention) is the verified target. `qwen3_5.py`'s `create_attention_mask(h, cache[fa_idx])` + `create_ssm_mask(h, cache[ssm_idx])` both go through the cache-aware `make_mask(N)` path, so masks are chunk-size-correct and SSM recurrent state carries across chunks via `ArraysCache.state` — exactly the safety profile you outlined.

On a ≈48K text-only prompt with `--prefill-step-size 512`, the peak single-buffer allocation for a 512-token chunk is `(1, num_heads, 512, 512) * 2 bytes` ≈ 40 MB per attention layer — orders of magnitude under the 72 GiB Metal single-buffer cap.

**Regression pin** (`TestVmlx89HybridChunkedPrefillOptIn` in `tests/test_vl_video_regression.py`) covers:
- Env var off (default) → hybrid text-only takes the full-model path.
- Env var on → hybrid text-only takes the chunked path with correct `prefill_step_size` chunking.
- Non-hybrid text-only → unchanged (chunked path regardless of env var).
- MLLM with images → unchanged (full-model path regardless of env var — VL preprocessing makes chunking unsafe).
- Env var value survives round-trip through server boot.

Ships in v1.3.63+. If you end up running Qwen3.6-35B-A3B (or any other hybrid) against a long-context workload with the opt-in set, I'd love to hear:
- Any perceivable quality diff vs the full-path
- Peak active-memory curve during prefill
- Whether `--kv-cache-quantization q8` interacts cleanly with the chunked path in your setup

Closing — reopen with your results if you want a specific hybrid architecture moved from opt-in to default-on.
